File: semgus-sketch/src/main/java/org/semgus/sketch/base/Stmt.java
package org.semgus.sketch.base;

import org.jetbrains.annotations.NotNull;

import java.util.List;
import java.util.stream.Stream;

public sealed interface Stmt extends Syntax {
  record VarDef(@NotNull Param decl, @NotNull Expr init) implements Stmt {}

  record FnDef(@NotNull Param decl, @NotNull List<Param> params, @NotNull Stmt body) implements Stmt {}

  record StructDef(@NotNull Id id, @NotNull List<Param> fields) implements Stmt {}

  record Atomic(@NotNull Id prefix, @NotNull Expr e) implements Stmt {}

  record Seq(@NotNull List<Stmt> ss) implements Stmt {}

  record Skip() implements Stmt {}

  static @NotNull Stmt varDef(@NotNull Param decl, @NotNull Expr init) {
    return new VarDef(decl, init);
  }

  static @NotNull Stmt fnDef(@NotNull Param decl, @NotNull List<Param> params, @NotNull Stmt body) {
    return new FnDef(decl, params, body);
  }

  static @NotNull Stmt structDef(@NotNull Id id, @NotNull List<Param> fields) {
    return new StructDef(id, fields);
  }

  static @NotNull Stmt ntTypeDef(@NotNull String ntName, @NotNull List<Param> fields) {
    return structDef(Id.id(ntName, IdAttr.withNTType()), fields);
  }

  static @NotNull Stmt atomic(@NotNull Id prefix, @NotNull Expr e) {
    return new Atomic(prefix, e);
  }

  static @NotNull Stmt aAssert(@NotNull Expr e) {
    return atomic(Id.keyAssert(), e);
  }

  static @NotNull Stmt bndChk() {
    var e = Expr.binaryPlain(">", Expr.refPlain("bnd"), Expr.refPlain(0));
    return Stmt.aAssert(e);
  }

  static @NotNull Stmt aReturn(@NotNull Expr e) {
    return atomic(Id.keyReturn(), e);
  }

  static @NotNull Stmt seq(@NotNull List<Stmt> ss) {
    return new Seq(ss);
  }

  static @NotNull Stmt bndChked(@NotNull List<Stmt> ss) {
    return seq(Stream.concat(Stream.of(bndChk()), ss.stream()).toList());
  }

  static @NotNull Stmt bndChked(@NotNull Stmt s) {
    return bndChked(List.of(s));
  }

  static @NotNull Stmt skip() {
    return new Skip();
  }
}

File: semgus-sketch/src/main/java/org/semgus/sketch/syntax/Stmt.kt
package org.semgus.sketch.syntax

/**
 * Sketch statement.
 */
internal sealed class Stmt : Syntax() {
  object Skip : Stmt()
  data class VarDef(val decl: Param, val init: Expr) : Stmt()
  data class FnDef(val decl: Param, val params: Sequence<Param>, val body: Stmt) : Stmt()
  data class StructDef(val id: Id, val fields: Sequence<Param>) : Stmt()
  data class Atomic(val prefix: Id, val e: Expr) : Stmt()
  data class Seq(val ss: Sequence<Stmt>) : Stmt()
}

internal fun skip() = Stmt.Skip

internal fun varDef(decl: Param, init: Expr) = Stmt.VarDef(decl, init)

internal fun bndDef(bnd: Long) = varDef(intPlain("bnd"), refPlain(bnd))

internal fun fnDef(decl: Param, params: Sequence<Param>, body: Stmt) = Stmt.FnDef(decl, params, body)

internal fun structDef(id: Id, fields: Sequence<Param>) = Stmt.StructDef(id, fields)
internal fun ntTypeDef(ntName: String, fieldParams: Sequence<Param>) =
  structDef(
    id = id(ntName) { withNTType() },
    fields = fieldParams.map { fieldParam ->
      param(fieldParam.type, id(fieldParam.id.name) { withField() })
    },
  )

internal fun atomic(prefix: Id, e: Expr) = Stmt.Atomic(prefix, e)
internal fun aAssert(e: Expr) = atomic(idPlain("assert"), e)
internal fun aBndChk() = aAssert(binary(Op.GT, refPlain("bnd"), refPlain(0)))
internal fun aReturn(e: Expr) = atomic(idPlain("return"), e)

internal fun seq(ss: Sequence<Stmt>) = Stmt.Seq(ss)
internal fun sBndChked(ss: Sequence<Stmt>) = seq(sequenceOf(aBndChk()) + ss)
internal fun sBndChked(s: Stmt) = seq(sequenceOf(aBndChk(), s))

internal fun sBndDefed(ss: Sequence<Stmt>, bnd: Long) = seq(sequenceOf(bndDef(bnd)) + ss)

internal fun sBndDefed(s: Stmt, bnd: Long) = seq(sequenceOf(bndDef(bnd), s))
